xla/backends/gpu/codegen/copy.cc

```diff
@@ -83,28 +83,57 @@ absl::StatusOr<FusionEmissionResult> DynamicMemcpyFusion::Emit(
     const HloFusionInstruction& fusion) const {
   CHECK_EQ(analysis_.fusion_roots().size(), 1);
 
-  const auto* src_instr = &analysis_.fusion_root(0).GetOperand(0).instruction();
+  auto root = analysis_.fusion_roots().front();
+
+  int source_operand_index;
+  const Shape* copy_shape;
+
+  if (root.opcode() == HloOpcode::kDynamicUpdateSlice) {
+    // We only handle in-place DUS operations (where the source and the
+    // destination are the same). This could be extended to out-of-place DUSes,
+    // but we would either have to issue two memcpys (one of the full original
+    // buffer, one for the updated slice), or three (one for the unchanged
+    // prefix, one for the updated slice, one for the unchanged suffix). The
+    // first option is inefficient, the second option is currently not
+    // implemented: we only support dynamic offsets, no dynamic sizes.
+    TF_ASSIGN_OR_RETURN(BufferAllocation::Slice input,
+                        buffer_assignment_->GetUniqueSlice(
+                            &root.GetOperand(0).instruction(), {}));
+    TF_ASSIGN_OR_RETURN(BufferAllocation::Slice dst,
+                        buffer_assignment_->GetUniqueSlice(&fusion, {}));
+    CHECK_EQ(input, dst);
+
+    source_operand_index = 1;
+    copy_shape = &root.GetOperand(source_operand_index).shape();
+  } else {
+    source_operand_index = 0;
```

Comment:
nit: could we assert on what the opcode is expected to be here? It's not 100% clear locally, though I expect it's a DS? :)

Reply:
Done.